src/crypto/rx-slow-hash.c

```diff
@@ -0,0 +1,349 @@
+}
+
+static char rx_althash[32];	// seedhash for alternate blocks
+
+void rx_alt_slowhash(const uint64_t mainheight, const uint64_t seedheight, const char *seedhash, const void *data, size_t length, char *hash) {
+  uint64_t s_height = me_rx_seedheight(mainheight);
+  int alt_toggle = (s_height & SEEDHASH_EPOCH_BLOCKS) == 0;
+  rx_state *rx_sp = &rx_s[alt_toggle];
+  if (rx_sp->rs_height != seedheight || rx_sp->rs_cache == NULL || memcmp(seedhash, rx_althash, sizeof(rx_althash))) {
+    memcpy(rx_althash, seedhash, sizeof(rx_althash));
+    rx_sp->rs_height = 1;
+    rx_seedhash_int(rx_sp, seedheight, seedhash, 0);
+  }
+  if (rx_vm == NULL) {
+    randomx_flags flags = RANDOMX_FLAG_DEFAULT;
+    if (use_rx_jit())
+      flags |= RANDOMX_FLAG_JIT;
+    if(!force_software_aes() && check_aes_hw())
+      flags |= RANDOMX_FLAG_HARD_AES;
+    rx_vm = randomx_create_vm(flags | RANDOMX_FLAG_LARGE_PAGES, rx_sp->rs_cache, NULL);
+    if(rx_vm == NULL) //large pages failed
+      rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, NULL);
+    if(rx_vm == NULL) {//fallback if everything fails
+      flags = RANDOMX_FLAG_DEFAULT;
+      rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, NULL);
+    }
+    if(rx_vm == NULL)
+      local_abort("Couldn't allocate RandomARQ VM");
+  }
+  randomx_calculate_hash(rx_vm, data, length, hash);
+}
+
+void me_rx_slow_hash(const void *data, size_t length, char *hash, int miners) {
+  if (rx_vm == NULL) {
+    rx_state *rx_sp = &rx_s[rx_s_toggle];
+    randomx_flags flags = RANDOMX_FLAG_DEFAULT;
+    if (use_rx_jit())
+      flags |= RANDOMX_FLAG_JIT;
+    if(!force_software_aes() && check_aes_hw())
+      flags |= RANDOMX_FLAG_HARD_AES;
+    if (miners) {
+      if (rx_dataset == NULL) {
+        CTHR_MUTEX_LOCK(rx_mutex);
+        if (rx_dataset == NULL) {
+          rx_dataset = randomx_alloc_dataset(RANDOMX_FLAG_LARGE_PAGES);
+          if (rx_dataset == NULL)
+            rx_dataset = randomx_alloc_dataset(RANDOMX_FLAG_DEFAULT);
+          if (rx_dataset != NULL)
+            rx_initdata(rx_sp->rs_cache, miners);
+        }
+        CTHR_MUTEX_UNLOCK(rx_mutex);
+      }
+      if (rx_dataset != NULL)
+        flags |= RANDOMX_FLAG_FULL_MEM;
+      else
+        miners = 0;
+    }
+    rx_vm = randomx_create_vm(flags | RANDOMX_FLAG_LARGE_PAGES, rx_sp->rs_cache, rx_dataset);
+    if(rx_vm == NULL) //large pages failed
+      rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, rx_dataset);
+    if(rx_vm == NULL) {//fallback if everything fails
+      flags = RANDOMX_FLAG_DEFAULT | (miners ? RANDOMX_FLAG_FULL_MEM : 0);
+      rx_vm = randomx_create_vm(flags, rx_sp->rs_cache, rx_dataset);
+    }
+    if (rx_vm == NULL)
+      local_abort("Couldn't allocate RandomARQ VM");
+  }
+  randomx_calculate_hash(rx_vm, data, length, hash);
+}
+
+void me_rx_slow_hash_allocate_state(void) {
+}
+
+void me_rx_slow_hash_free_state(void) {
+  if (rx_vm != NULL) {
+    randomx_destroy_vm(rx_vm);
+    rx_vm = NULL;
+  }
+}
```

Comment:
That rx-slow-hash.c does not match ours :)

Reply:
removed